src/ir/module-utils.h

```diff
@@ -328,6 +308,58 @@ template<typename T> struct CallGraphPropertyAnalysis {
 
     PassRunner runner(&wasm);
     Mapper(&wasm, &map, work).run(&runner, &wasm);
+  }
+};
+
+// Helper class for analyzing the call graph.
+//
+// Provides hooks for running some initial calculation on each function (which
+// is done in parallel), writing to a FunctionInfo structure for each function.
+// Then you can call propagateBack() to propagate a property of interest to the
+// calling functions, transitively.
+//
+// For example, if some functions are known to call an import "foo", then you
+// can use this to find which functions call something that might eventually
+// reach foo, by initially marking the direct callers as "calling foo" and
+// propagating that backwards.
+template<typename T> struct CallGraphPropertyAnalysis {
+  Module& wasm;
+
+  // The basic information for each function about whom it calls and who is
+  // called by it.
+  struct FunctionInfo {
+    std::set<Function*> callsTo;
+    std::set<Function*> calledBy;
+  };
+
+  typedef std::map<Function*, T> Map;
+  Map map;
+
+  typedef std::function<void(Function*, T&)> Func;
+
+  CallGraphPropertyAnalysis(Module& wasm, Func work) : wasm(wasm) {
+    ParallelFunctionAnalysis<T> analysis(wasm, [&](Function* func, T& info) {
```

Comment:
Can `T` just be `FunctionInfo` here?

Reply:
No, as the actual T may include other fields (and does in general, see Asyncify, where we store stuff like whether a function can sleep).